src/blockstorage/prune.cpp

```diff
@@ -0,0 +1,418 @@
+    }
+    return retval;
+}
+
+void PruneFiles(std::set<int> &setFilesToPrune, uint64_t nLastBlockWeCanPrune)
+{
+    uint64_t nCurrentUsage = CalculateCurrentUsage();
+    // We don't check to prune until after we've allocated new space for files
+    // So we should leave a buffer under our target to account for another allocation
+    // before the next pruning.
+    uint64_t nBuffer = blockfile_chunk_size + undofile_chunk_size;
+    uint64_t nBytesToPrune;
+    int count = 0;
+
+    if (nCurrentUsage + nBuffer >= nPruneTarget)
+    {
+        for (int fileNumber = 0; fileNumber < nLastBlockFile; fileNumber++)
+        {
+            nBytesToPrune = vinfoBlockFile[fileNumber].nSize + vinfoBlockFile[fileNumber].nUndoSize;
+
+            if (vinfoBlockFile[fileNumber].nSize == 0)
+            {
+                continue;
+            }
+
+            if (nCurrentUsage + nBuffer < nPruneTarget) // are we below our target?
+            {
+                break;
+            }
+
+            // don't prune files that could have a block within MIN_BLOCKS_TO_KEEP of the main chain's tip but keep
+            // scanning
+            if (vinfoBlockFile[fileNumber].nHeightLast > nLastBlockWeCanPrune)
+            {
+                continue;
+            }
+
+            PruneOneBlockFile(fileNumber);
+            // Queue up the files for removal
+            setFilesToPrune.insert(fileNumber);
+            nCurrentUsage -= nBytesToPrune;
+            count++;
+        }
+    }
+
+    LOG(PRUNE, "Prune: target=%dMiB actual=%dMiB diff=%dMiB max_prune_height=%d removed %d blk/rev pairs\n",
+        nPruneTarget / 1024 / 1024, nCurrentUsage / 1024 / 1024,
+        ((int64_t)nPruneTarget - (int64_t)nCurrentUsage) / 1024 / 1024, nLastBlockWeCanPrune, count);
+}
+
+uint64_t PruneDB(uint64_t nLastBlockWeCanPrune)
+{
+    CBlockIndex *pindexOldest = chainActive.Tip();
+    while (pindexOldest->pprev)
+    {
+        pindexOldest = pindexOldest->pprev;
+    }
+    uint64_t prunedCount = 0;
+    std::vector<std::string> blockBatch;
+    std::vector<std::string> undoBatch;
+    while (pindexOldest != nullptr)
+    {
+        if (pindexOldest->GetBlockHash() == Params().GetConsensus().hashGenesisBlock)
+        {
+            pindexOldest = chainActive.Next(pindexOldest);
+            continue;
+        }
+        if (pindexOldest->nFile == 0)
+        {
+            // already pruned block
+            pindexOldest = chainActive.Next(pindexOldest);
+            continue;
+        }
+        if (!fPruneWithMask && nDBUsedSpace < nPruneTarget)
+        {
+            break;
+        }
+        if (pindexOldest->nHeight >= (int)nLastBlockWeCanPrune)
+        {
+            break;
+        }
+        if (fPruneWithMask == true && hashMaskCompare(pindexOldest->GetBlockHash()))
+        {
+            pindexOldest = chainActive.Next(pindexOldest);
+            continue;
+        }
+        unsigned int blockSize = pindexOldest->nDataPos;
```

Comment:
Add comment here:
// We've chosen to prune this block

Reply:
added